docs(combobox): improve Combobox docs (#2714)
twilio-labs/paste: packages/paste-website/src/pages/components/combobox/index.mdx

The Combobox is an opinionated implementation and wrapper around the **very excellent [Downshift component](https://downshift.netlify.app/downshift)**.

> Combobox allows a user to make a selection from a styled list box of options. Each option can consist of more than just text,

Reviewer: ```suggestion
Combobox allows a user to make a selection from a styled listbox of options. Each option can consist of more than just text,
```

_e.g._ text paired with an icon. It can also be set up with autocomplete functionality so users can easily find a

> specific option.

Reviewer: Change to this from the spec:

`A Combobox allows a user to select a single value from a styled listbox of options. Each option can consist of more than just text, _e.g._ text paired with an icon. However, an option can only be paired with non-interactive content and shouldn’t contain components like a checkbox or button. It can also be set up with autocomplete functionality so users can easily find a specific option.`

> specific option.

Reviewer: todo: Change to this from the spec:

todo: Change to this from the spec:

`A Combobox allows a user to select a single value from a styled listbox of options. Each option can consist of more than just text, _e.g._ text paired with an icon. However, an option can only be paired with non-interactive content and shouldn’t contain components like a checkbox or button. It can also be set up with autocomplete functionality so users can easily find a specific option.

This component is an opinionated implementation and wrapper around the [Downshift component](https://downshift.netlify.app/downshift).`

### What’s the difference between Select and Combobox?

At its most basic, a Select has an options list that’s styled according to the browser default.
A Combobox has a Twilio-styled options list and can allow for additional functionality like
autocomplete and grouping options.

Use a [Select](/components/select) when:

- You need a native picker experience, especially on mobile devices.
- Users will be selecting from a list of 4-10 options, or a sorted list of highly familiar options (e.g., alphabetical list of states or countries).

> - You need the component to work out-of-the-box across all operating systems and browsers.

Reviewer: todo: From the spec, change to:

```
- You need a native picker experience, especially on mobile devices.
- You only need to show text in an option.
- Users will be selecting from a list of 4-10 options, or a sorted list of highly familiar options (e.g., alphabetical list of states or countries). For less than 4 items, consider using a [Radio Group](https://paste.twilio.design/components/radio-group) instead.
- You need the component to work out-of-the-box across all operating systems and browsers.
```

> Use a [Combobox](/components/combobox) when:

Reviewer: ```suggestion
Use a Combobox when:
```

- You need a Twilio-styled options list.
- You need to show more than text in an option (e.g., text paired with an icon).
- You need to group options under labels.
- You need to disable options in the list.
- Users would benefit from autocomplete functionality (e.g., autocomplete, search). For example, autocomplete may be useful when users need to select from a list of more than 10 options.

> - You need to lazy load a much longer list of options to improve page load performance.

Reviewer: todo: From the spec, change to:

```
- You need a Twilio-styled options list.
- You need to show more than text in an option (e.g., text paired with an icon).
- You need to group options under labels.
- You need to disable options in the list.
- You need placeholder text in the field to show an example option, rather than a preselected option.
- Users would benefit from autocomplete functionality. For example, autocomplete may be useful when users need to select from a list of more than 10 options.
- You need to lazy load a much longer list of options to improve page load performance.
```

### Accessibility

Combobox is built with consideration for the [ARIA combobox pattern](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/).

When a user is focused on a Combobox, the listbox opens. When a user makes a selection, the listbox closes so the selected option can be registered to screen readers.

#### Keyboard interaction

When the user is focused on a Combobox, the following keyboard interactions apply:

- Up and down arrows move the user between the options
- Enter selects the currently active option

> 

Reviewer: todo: Add to the bottom of the list:

`- Escape closes the listbox or clears the selected option if the listbox is already closed`

## Examples

### Basic Combobox

Use a basic Combobox to allow users to select a value from a list of
predefined options.

<LivePreview scope={{Combobox}} noInline language="jsx" showOverflow>
  {defaultExample}
</LivePreview>

> 

Reviewer: todo: Add a "Combobox with pre-selected value" example based on this section of the spec: https://docs.google.com/document/d/143idKgnfqFCmSceZZ67nIcLCxRnBsuxYPB8TMxTxKno/edit#bookmark=id.ccmpxz9yjlvr

Author: The basic combobox has a preselected value to simplify this page. Do you still want it standalone?

Reviewer: ahh, good point. Ignore this comment.

### Autocomplete Combobox

The autocomplete Combobox allows users to filter the list of options by typing into the input directly.

> 

Reviewer: todo: Change to:

```
The autocomplete Combobox allows users to filter the list of options by typing into the input directly.

This is useful when a user needs to filter a list of options, or when there are more than 15 options as users may find it difficult to navigate with only scrolling. Pass the `autocomplete` prop to enable the autocomplete feature.
```

  {groupedLabelComboboxExample}
</LivePreview>

### Combobox using Option Template

Use the option template to display more complex options in a listbox.

The `optionTemplate` prop allows you to pass `jsx` to customize the options. Note that we use native HTML input elements
to build Combobox and these HTML elements don't allow for images, icons, or svgs to be added even with the option template.

<LivePreview
  scope={{Combobox, LinkExternalIcon, MediaBody, MediaObject, MediaFigure, Text}}
  noInline
  language="jsx"
  showOverflow
>

>   {optionTemplateExample}

Reviewer: Since having links/anchors as combobox options is kinda jank, can you remove the ExternalLinkIcon from the optionTemplateExample?

It can look something like this: https://docs.google.com/document/d/143idKgnfqFCmSceZZ67nIcLCxRnBsuxYPB8TMxTxKno/edit#bookmark=id.aa46w1hfargd

  {controlledComboboxExample}
</LivePreview>

> ### useCombobox state hook

Reviewer: Same comment here as above about removing the ExternalLinkIcon from stateHookCombobox

<Callout variant="warning" marginY="space70">

>   <CalloutHeading as="h5">Power user move!</CalloutHeading>

Reviewer: nitpick: There's a sentence hidden here that says "An example usecase..." where it should be "An example use case..."

## States

### Disabled Combobox

<LivePreview scope={{Combobox}} noInline language="jsx" showOverflow>
  {disabledExample}
</LivePreview>

### Combobox with disabled options

<LivePreview scope={{Combobox}} noInline language="jsx" showOverflow>

>   {disabledOptionsExample}

Reviewer: todo(blocking): This example needs a disabled option.

</LivePreview>

### Combobox with inline error

<LivePreview scope={{Combobox}} noInline language="jsx" showOverflow>
  {errorExample}
</LivePreview>

### Combobox with an empty state

Use an empty state to indicate to a user that their input does not match any value in the list of options.

<LivePreview scope={{Combobox, Box, Text}} noInline language="jsx" showOverflow>

>   {emptyStateExample}

Reviewer: todo: This example needs an empty state

</LivePreview>

## Composition Notes

A Combobox is comprised of a label, an input and a listbox.

### Positioning form fields

Stack form fields vertically with `$space-80` spacing between each field. Avoid placing multiple form fields on the

> same horizontal row to help make it easier to scan a page vertically. Use the Grid component if you need to place
> form fields horizontally.

Reviewer: ```suggestion
same horizontal row to help make it easier to scan a page vertically.
```

### Options in a Combobox

Keep option text concise and simple. Option text will truncate when it is longer than the width of the container.

Use at least five options in a Combobox. If there are less than five options or if choosing options
requires more explanation, consider using a Checkbox instead and add help text for each option,
or give more explanation through help text. Sort options logically (e.g., alphabetically, by value)

> or in an order that’s intuitive to your user.

Reviewer: todo: Change to this from the spec:

```
Keep option text concise and simple. 

If setting a default selected value, use a safe and reversible option as the default selected value.

Use at least seven options in a Combobox. If there are less than seven options or if choosing options requires more explanation, consider using a [Radio Group](/components/radio-group) instead and add help text for each option, or give more explanation through help text. Sort options logically (e.g., alphabetically, by value) or in an order that’s intuitive to your user.
```

| `onSelectedItemChange?`     | `(newSelectedItem: any[]) => void`                    | Callback function for after an item is selected or deselected.                                                                                            |                |
| `onHighlightedIndexChange?` | `(newHighlightedIndex: number) => void`               | Callback function for after an item is highlighted. Items are highlighted if they are hovered over or with keyboard actions.                              |                |
| `onInputValueChange?`       | `(newInputValue: string) => void`                     | Callback function for after the input value is changed.                                                                                                   |                |
| `onIsOpenChange?`           | `(newIsOpen: boolean) => void`                        | Callback function for after the Combobox is opened or closed.                                                                                             |                |

> | `state?`                    | `UseComboboxPrimitiveReturnValue<any>`                | Used with the `useCombobox` hook. When using this prop, all other state props are **ignored**. They must be passed to `useCombobox` as arguments instead. |                |

Author: Completely agree. We tried to add this a couple years ago but couldn't get it to work the way we wanted, and have kicked the can on it so long we simply forgot to try again after so long. Have you used that tool before? It's worth making a ticket to investigate if you're interested!

#### useCombobox arguments

> 

Reviewer: Can you add a Figma section (as an H3) right before the Changelog, with this content?

https://docs.google.com/document/d/143idKgnfqFCmSceZZ67nIcLCxRnBsuxYPB8TMxTxKno/edit#bookmark=kix.ih4ca8a5a8co
